File: stations/hook-dispatcher/internal/handlers/read_edit_ratio.go
package handlers

// read_edit_ratio.go — Go port of handlers/read_edit_ratio.py
//
// PostToolUse / Read|Edit|Write: tracks Read vs Edit/Write calls per session.
// When the model edits files without reading them first (blind-edit anti-pattern),
// emits a warning message and fires background notifications.
//
// Thresholds (from empirical data):
//   - Good:    read:edit ratio >= 4.0, blind edit rate < 10%
//   - Warning: ratio < 3.0 OR blind edit rate > 25%
//   - Alert:   ratio < 2.0 OR blind edit rate > 35%
//
// State: /tmp/.read-edit-ratio-{hash12}.json (per-session)

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"time"

	"github.com/joneshong/hook-dispatcher/internal/core"
	portregistry "github.com/joneshong/workshop/libs/go-port-registry"
)

// rerBarkServer points at the bark notification daemon (port 8090 in the
// registry). Resolved at init so yaml is single source of truth.
var rerBarkServer = portregistry.URL("bark", "", 8090)

const (
	rerBarkDeviceKey = "gx7KnK5f8iAKuqNLWzy5hP"

	rerMinEditsForWarn = 5
	rerRatioWarn       = 3.0
	rerRatioAlert      = 2.0
	rerBlindWarn       = 0.25
	rerBlindAlert      = 0.35
	rerWarnCooldownS   = 300 // 5 min
)

// NOTE: Not registered. Python handlers/__init__.py does not import or register
// read_edit_ratio — it's inactive in production. Keeping the implementation for
// future opt-in, but no core.Register() call to preserve parity.
func init() {}

// rerState mirrors the JSON fields stored in /tmp/.read-edit-ratio-<hash>.json
type rerState struct {
	Reads      []string `json:"reads"`
	ReadCount  int      `json:"read_count"`
	EditCount  int      `json:"edit_count"`
	BlindEdits int      `json:"blind_edits"`
	LastWarnAt float64  `json:"last_warn_at"`
	StartedAt  float64  `json:"started_at"`
}

func rerStatePath(sessionID string) string {
	h := sha256.Sum256([]byte(sessionID))
	hex12 := fmt.Sprintf("%x", h[:6]) // 12 hex chars
	return filepath.Join(os.TempDir(), fmt.Sprintf(".read-edit-ratio-%s.json", hex12))
}

func rerLoad(path string) rerState {
	data, err := os.ReadFile(path)
	if err != nil {
		return rerState{}
	}
	var s rerState
	if err := json.Unmarshal(data, &s); err != nil {
		return rerState{}
	}
	return s
}

func rerSave(path string, s rerState) {
	data, err := json.Marshal(s)
	if err != nil {
		return
	}
	_ = os.WriteFile(path, data, 0o644)
}

func rerExtractSessionID(rawInput string) string {
	var parsed map[string]any
	if err := json.Unmarshal([]byte(rawInput), &parsed); err != nil {
		return "default"
	}
	if v, ok := parsed["session_id"].(string); ok && v != "" {
		return v
	}
	return "default"
}

func rerExtractFilePath(toolInput map[string]any) string {
	if v, ok := toolInput["file_path"].(string); ok {
		return v
	}
	return ""
}

func readEditRatioHandle(_, toolName string, toolInput map[string]any, rawInput string) core.HookResult {
	sessionID := rerExtractSessionID(rawInput)
	sp := rerStatePath(sessionID)
	state := rerLoad(sp)

	if state.ReadCount == 0 && state.EditCount == 0 && state.StartedAt == 0 {
		state = rerState{
			Reads:     []string{},
			StartedAt: float64(time.Now().UnixNano()) / 1e9,
		}
	}

	filePath := rerExtractFilePath(toolInput)

	if toolName == "Read" {
		state.ReadCount++
		if filePath != "" && !rerContains(state.Reads, filePath) {
			state.Reads = append(state.Reads, filePath)
			if len(state.Reads) > 500 {
				state.Reads = state.Reads[len(state.Reads)-500:]
			}
		}
		rerSave(sp, state)
		return core.Allow()
	}

	// Edit or Write
	state.EditCount++
	if filePath != "" && !rerContains(state.Reads, filePath) {
		state.BlindEdits++
	}

	editCount := state.EditCount
	readCount := state.ReadCount
	blindEdits := state.BlindEdits
	rerSave(sp, state)

	// Not enough data yet
	if editCount < rerMinEditsForWarn {
		return core.Allow()
	}

	ratio := float64(readCount) / float64(editCount)
	blindRate := float64(blindEdits) / float64(editCount)

	// Check cooldown
	now := float64(time.Now().UnixNano()) / 1e9
	if now-state.LastWarnAt < rerWarnCooldownS {
		return core.Allow()
	}

	isAlert := ratio < rerRatioAlert || blindRate > rerBlindAlert
	isWarn := ratio < rerRatioWarn || blindRate > rerBlindWarn

	if !isAlert && !isWarn {
		return core.Allow()
	}

	// Update cooldown
	state.LastWarnAt = now
	rerSave(sp, state)

	severity := "🟡"
	if isAlert {
		severity = "🔴"
	}

	msg := fmt.Sprintf(
		"%s [read:edit ratio] R:E=%.1f (目標≥4.0), 盲改率=%.0f%% (%d/%d), reads=%d edits=%d",
		severity, ratio, blindRate*100, blindEdits, editCount, readCount, editCount,
	)
	if isAlert {
		msg += " — 模型可能進入淺思考模式, 建議重啟 session"
	}

	// Fire-and-forget notifications
	rerNotify(severity, ratio, blindRate, isAlert)

	return core.Message(msg)
}

func rerContains(slice []string, s string) bool {
	for _, v := range slice {
		if v == s {
			return true
		}
	}
	return false
}

func rerNotify(severity string, ratio float64, blindRate float64, isAlert bool) {
	title := severity + " Read:Edit Ratio"
	body := fmt.Sprintf("R:E=%.1f, 盲改率=%.0f%%", ratio, blindRate*100)
	if isAlert {
		body += " — 建議重啟 session"
	}

	// macOS notification
	osaBody := rerEscapeOSA(body)
	osaTitle := rerEscapeOSA(title)
	osaCmd := fmt.Sprintf(
		`osascript -e 'display notification "%s" with title "%s" sound name "Sosumi"'`,
		osaBody, osaTitle,
	)
	_ = core.RunBackground([]string{"sh", "-c", osaCmd}, "")

	// Bark push
	sound := "bell"
	level := "active"
	if isAlert {
		sound = "alarm"
		level = "timeSensitive"
	}
	barkURL := fmt.Sprintf(
		"%s/%s/%s/%s?group=hook-observatory&sound=%s&level=%s",
		rerBarkServer, rerBarkDeviceKey,
		url.PathEscape(title), url.PathEscape(body),
		sound, level,
	)
	_ = core.RunBackground([]string{"curl", "-sf", barkURL}, "")
}

func rerEscapeOSA(s string) string {
	// Escape double-quotes and backslashes for osascript string literals
	out := ""
	for _, ch := range s {
		switch ch {
		case '"':
			out += `\"`
		case '\\':
			out += `\\`
		default:
			out += string(ch)
		}
	}
	return out
}

File: stations/_archive/hook-observatory-py/handlers/read_edit_ratio.py
"""
Read:Edit ratio monitor — detects blind-edit anti-pattern.

Tracks Read vs Edit/Write tool calls per session. When the model edits
files without reading them first, it signals shallow reasoning (the
"editing without reading" pattern identified in Laurenzo's 6,852-session
analysis).

Thresholds (from empirical data):
  - Good: read:edit ratio >= 4.0, blind edit rate < 10%
  - Warning: ratio < 3.0 OR blind edit rate > 25%
  - Alert: ratio < 2.0 OR blind edit rate > 35%

State: /tmp/.read-edit-ratio-{hash}.json (per-session, auto-cleaned)
Latency: <1ms (JSON read/write, no network)
"""

from __future__ import annotations

import hashlib
import json
import os
import tempfile
import time

from .base import ALLOW, HookResult, message, run_background

# --- Bark config ---
_BARK_SERVER = "http://localhost:8090"
_BARK_DEVICE_KEY = "gx7KnK5f8iAKuqNLWzy5hP"

# --- Thresholds ---
_MIN_EDITS_FOR_WARN = 5  # Don't warn until enough data
_RATIO_WARN = 3.0  # read:edit ratio below this → warn
_RATIO_ALERT = 2.0  # below this → strong alert
_BLIND_WARN = 0.25  # blind edit rate above 25% → warn
_BLIND_ALERT = 0.35  # above 35% → strong alert
_WARN_COOLDOWN_S = 300  # 5 min between warnings


def _state_path(session_id: str) -> str:
    h = hashlib.sha256(session_id.encode()).hexdigest()[:12]
    return os.path.join(tempfile.gettempdir(), f".read-edit-ratio-{h}.json")


def _load(path: str) -> dict:
    try:
        with open(path) as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError):
        return {}


def _save(path: str, state: dict) -> None:
    try:
        with open(path, "w") as f:
            json.dump(state, f, separators=(",", ":"))
    except OSError:
        pass


def _extract_session_id(raw_input: str) -> str:
    try:
        parsed = json.loads(raw_input)
        return str(parsed.get("session_id", "")) or "default"
    except (json.JSONDecodeError, AttributeError):
        return "default"


def _extract_file_path(tool_input: dict) -> str:
    return tool_input.get("file_path", "")


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    if event_type != "PostToolUse":
        return ALLOW

    # Only track Read, Edit, Write
    if tool_name not in ("Read", "Edit", "Write"):
        return ALLOW

    session_id = _extract_session_id(raw_input)
    sp = _state_path(session_id)
    state = _load(sp)

    if not state:
        state = {
            "reads": [],  # file paths that have been Read
            "read_count": 0,
            "edit_count": 0,
            "blind_edits": 0,
            "last_warn_at": 0,
            "started_at": time.time(),
        }

    file_path = _extract_file_path(tool_input)

    if tool_name == "Read":
        state["read_count"] = state.get("read_count", 0) + 1
        reads = state.get("reads", [])
        if file_path and file_path not in reads:
            reads.append(file_path)
            # Cap list size to avoid unbounded growth
            if len(reads) > 500:
                reads = reads[-500:]
        state["reads"] = reads
        _save(sp, state)
        return ALLOW

    # Edit or Write
    state["edit_count"] = state.get("edit_count", 0) + 1

    # Check if file was previously Read in this session
    reads = state.get("reads", [])
    if file_path and file_path not in reads:
        state["blind_edits"] = state.get("blind_edits", 0) + 1

    edit_count = state["edit_count"]
    read_count = state.get("read_count", 0)
    blind_edits = state.get("blind_edits", 0)

    _save(sp, state)

    # Not enough data yet
    if edit_count < _MIN_EDITS_FOR_WARN:
        return ALLOW

    # Calculate metrics
    ratio = read_count / edit_count if edit_count > 0 else float("inf")
    blind_rate = blind_edits / edit_count if edit_count > 0 else 0

    # Check cooldown
    now = time.time()
    if now - state.get("last_warn_at", 0) < _WARN_COOLDOWN_S:
        return ALLOW

    # Determine severity
    alert = False
    warn = False

    if ratio < _RATIO_ALERT or blind_rate > _BLIND_ALERT:
        alert = True
    elif ratio < _RATIO_WARN or blind_rate > _BLIND_WARN:
        warn = True

    if not alert and not warn:
        return ALLOW

    # Update cooldown
    state["last_warn_at"] = now
    _save(sp, state)

    severity = "🔴" if alert else "🟡"
    msg = (
        f"{severity} [read:edit ratio] "
        f"R:E={ratio:.1f} (目標≥4.0), "
        f"盲改率={blind_rate:.0%} ({blind_edits}/{edit_count}), "
        f"reads={read_count} edits={edit_count}"
    )

    if alert:
        msg += " — 模型可能進入淺思考模式, 建議重啟 session"

    # Push notifications (fire-and-forget, non-blocking)
    _notify(severity, ratio, blind_rate, alert)

    return message(msg)


def _notify(severity: str, ratio: float, blind_rate: float, is_alert: bool) -> None:
    """Send macOS notification + Bark push in background."""
    from urllib.parse import quote

    title = f"{severity} Read:Edit Ratio"
    body = f"R:E={ratio:.1f}, 盲改率={blind_rate:.0%}"
    if is_alert:
        body += " — 建議重啟 session"

    # macOS notification center
    osa_body = body.replace('"', '\\"')
    osa_title = title.replace('"', '\\"')
    run_background(
        f'osascript -e \'display notification "{osa_body}" '
        f'with title "{osa_title}" sound name "Sosumi"\''
    )

    # Bark push
    bark_url = (
        f"{_BARK_SERVER}/{_BARK_DEVICE_KEY}"
        f"/{quote(title)}/{quote(body)}"
        f"?group=hook-observatory&sound={'alarm' if is_alert else 'bell'}"
        f"&level={'timeSensitive' if is_alert else 'active'}"
    )
    run_background(["curl", "-sf", bark_url])
